base/reflection.jl

```diff
@@ -440,6 +434,31 @@ end
 
 type_alignment(x::DataType) = (@_pure_meta; ccall(:jl_get_alignment, Csize_t, (Any,), x))
 
+"""
+    fieldcount(t::Type)
+
+Get the number of fields that an instance of the given type would have.
+An error is thrown if the type is too abstract to determine this.
+"""
+function fieldcount(t::ANY)
+    if t isa UnionAll || t isa Union
+        p = ccall(:jl_argument_datatype, Ptr{Void}, (Any,), t)
```

Comment:
would be cleaner to return the Julia sentinel (`nothing`) rather than the C sentinel